File: KantaroBLL/Distanze/Distanza.vb
<DebuggerDisplay("{GetText}")>
Public Class Distanza

    Private _valore As Single ' in semitoni

    ''' <summary>
    '''  Valore che indica la distanza in Toni
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property Toni As Single
        Get
            Return _valore / 2.0!
        End Get
    End Property

    ''' <summary>
    ''' Valore che indica la distanza in semitoni
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Property Semitoni As Integer
        Get
            Return _valore
        End Get
        Set(value As Integer)
            _valore = value
        End Set
    End Property

    Private ReadOnly Property GetText() As String
        Get
            Return String.Format("{0}", Me.Toni)
        End Get
    End Property

    ''' <summary>
    ''' Rappresenta il valore della distanza espresso in Toni
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Property Valore As Single
        Get
            Return _valore / 2.0!
        End Get
        Set(value As Single)
            If (2 * value) Mod 1 <> 0 Then
                Throw New ArgumentOutOfRangeException("Impossibile convertire il valore in Toni e Semitoni")
            End If
            _valore = value * 2.0!
        End Set
    End Property

    Public Shared Operator +(ByVal class1 As Distanza, ByVal class2 As Distanza) As Distanza
        Dim r = New Distanza()
        r.Valore = class1.Valore + class2.Valore
        Return r
    End Operator

    Public Shared Operator +(ByVal class1 As Distanza, ByVal valore As Integer) As Distanza
        Dim r = New Distanza()
        r.Valore = class1.Valore + valore
        Return r
    End Operator

    Public Shared Operator +(ByVal class1 As Distanza, ByVal valore As Double) As Distanza
        Dim r = New Distanza()
        r.Valore = class1.Valore + CSng(valore)
        Return r
    End Operator

    Public Shared Operator *(ByVal class1 As Distanza, ByVal moltiplicatore As Integer) As Distanza
        Dim d As New Distanza
        d.Valore = class1.Valore * moltiplicatore
        Return d
    End Operator

    Public Shared Operator *(ByVal moltiplicatore As Integer, ByVal class1 As Distanza) As Distanza
        Dim d As New Distanza
        d.Valore = moltiplicatore * class1.Valore
        Return d
    End Operator

    Public Shared Operator =(ByVal class1 As Distanza, ByVal class2 As Distanza) As Boolean
        Return class1.Equals(class2)
    End Operator

    Public Shared Operator <>(ByVal class1 As Distanza, ByVal class2 As Distanza) As Boolean
        Return Not class1.Equals(class2)
    End Operator

    Public Overrides Function Equals(obj As Object) As Boolean
        If Not TypeOf obj Is Distanza Then Return False
        Return DirectCast(obj, Distanza).Valore.Equals(Valore)
    End Function

    Public Overloads Shared Function Equals(obj1 As Object, obj2 As Object) As Boolean
        If Not TypeOf obj1 Is Distanza Then Return False
        Return DirectCast(obj1, Distanza).Equals(obj2)
    End Function

    Public Overrides Function ToString() As String
        Return Me.Valore.ToString()
    End Function

    Public Shared Narrowing Operator CType(ByVal initialData As Double) As Distanza
        Return CType(CSng(initialData), Distanza)
    End Operator

    Public Shared Narrowing Operator CType(ByVal initialData As Integer) As Distanza
        Return CType(CSng(initialData), Distanza)
    End Operator

    Public Shared Narrowing Operator CType(ByVal initialData As Single) As Distanza
        Dim r = New Distanza()
        Try
            r.Valore = initialData
        Catch ex As ArgumentOutOfRangeException
            Throw New InvalidCastException("Impossibile convertire il valore in un oggetto di tipo 'Distanza'")
        End Try
        Return r
    End Operator

    Public Shared Widening Operator CType(ByVal initialData As Distanza) As Single
        Return initialData.Valore
    End Operator

    Public Shared Narrowing Operator CType(ByVal initialData As String) As Distanza
        Return CType(Single.Parse(initialData), Distanza)
    End Operator

    Public Shared Widening Operator CType(ByVal initialData As Distanza) As String
        Return initialData.Valore.ToString()
    End Operator

End Class

File: Enzo.Music/Distanze/Distanza.cs
using System.Diagnostics;
using System.Numerics;

namespace Enzo.Music;

[DebuggerDisplay("{GetText}")]
public class Distanza : IAdditionOperators<Distanza, Distanza, Distanza>, IAdditionOperators<Distanza, int, Distanza>, IAdditionOperators<Distanza, double, Distanza>
    , IMultiplyOperators<Distanza, int, Distanza>
    , IEqualityOperators<Distanza, Distanza, bool>
{
    /// <summary>
    /// Valore che indica la distanza in Toni
    /// </summary>
    public float Toni => (float)Semitoni/2;

    /// <summary>
    /// Valore che indica la distanza in semitoni
    /// </summary>
    public int Semitoni { get; set; }

    /// <summary>
    /// Rappresenta il valore della distanza espresso in Toni
    /// </summary>
    public float Valore
    {
        get => (float)Semitoni / 2;
        set
        {
            if (2 * value % 1 != 0) // Consente la conversione solo per i valori interi o con il ".5"
            {
                throw new ArgumentOutOfRangeException(nameof(value), "Impossibile convertire il valore in Toni e Semitoni");
            }
            Semitoni =(int)(value * 2);
        }
    }

    private string GetText => Toni.ToString();

    public static Distanza operator +(Distanza left, Distanza right)
    {
        return new Distanza() { Valore = left.Valore + right.Valore };
    }

    public static Distanza operator +(Distanza left, int right)
    {
        return new Distanza() { Valore = left.Valore + right };
    }

    public static Distanza operator +(Distanza left, double right)
    {
        return new Distanza() { Valore = left.Valore + (float)right };
    }

    public static Distanza operator *(Distanza left, int right)
    {
        return new Distanza() { Valore = left.Valore * right };
    }
    public static Distanza operator *(int left, Distanza right)
    {
        return right * left;
    }

    public static bool operator ==(Distanza? left, Distanza? right)
    {
        return left?.Equals(right) ?? false;
    }

    public static bool operator !=(Distanza? left, Distanza? right)
    {
        return !(left==right);
    }

    public override bool Equals(object? obj)
    {
        if (obj is null || obj is not Distanza) return false;
        return (obj as Distanza)?.Valore.Equals(Valore) ?? false;
    }
    public static bool Equals(object obj, Distanza? other)
    {
        if (obj is null || obj is not Distanza) return false;
        return (obj as Distanza)?.Equals(other) ?? false;
    }

    public static explicit operator Distanza(double initialData)
    {
        return (Distanza)Convert.ToSingle(initialData);
    }

    public static explicit operator Distanza(int initialData)
    {
        return (Distanza)Convert.ToSingle(initialData);
    }

    public static explicit operator Distanza(float initialData)
    {
        var r = new Distanza();
        try
        {
            r.Valore = initialData;
        }
        catch //(ArgumentOutOfRangeException ex)
        {
            throw new InvalidCastException("Impossibile convertire il valore in un oggetto di tipo 'Distanza'");
        }
        return r;
    }

    public static implicit operator float(Distanza initialData)
    {
        return initialData.Valore;
    }

    public static explicit operator Distanza(string initialData)
    {
        return (Distanza)float.Parse(initialData);
    }

    public static implicit operator string(Distanza initialData)
    {
        return initialData.Valore.ToString();
    }

    public override string ToString()
    {
        return Valore.ToString();
    }

    public override int GetHashCode()
    {
        return Semitoni.GetHashCode();
    }
}
